### PR TITLE
fix(Dockerfile): readability and golang image update

### PR DESCRIPTION
**Pull request by mridulganga**
Summary of changes
1. Made changes to Dockerfile to use golang's alpine image for smaller size which helps build image faster
2. Added `make build` to build the golang binary replacing the embedded build command
3. Added an Optional PORT env var and expose command
4. Minor changes to Makefile to use specific main.go file instead of folder, and added few build args (-v - listing packages while building and -a - rebuilding prebuilt packages everytime) and envs (to disable cgo and setting GOOS)
5. Minor change in sonar-project.properties to add test inclusions parameter.
6. Minor change in .gitignore to ignore vs code and pycharm IDE config folders.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,13 +1,27 @@
...
+
 # numeric version of user nonroot:nonroot provided in image
 USER 65532:65532
+
+# run the executable
 CMD ["/app"]
+
+# optionally expose a port
+ENV PORT=8000
+EXPOSE $PORT
```
Comment (haani-niyaz) [inline Dockerfile:27]: I'm on the fence about this one for the reasons covered earlier. To reiterate not all projects will expose a service e.g: queue workers, CLIs etc. I'm of the opinion we should enable teams to add things and not have them worry about removing things.

File: Makefile [hunk 2]
```diff
@@ -37,4 +39,4 @@ fmt:
 ## build: Build binary into bin/ directory
 .PHONY: build
 build:
-	go build -ldflags="-w -s" -o bin/app ./...
+	$(GOFLAGS) $(GO_BUILD) -a -v -ldflags="-w -s" -o bin/app cmd/main.go
```
Comment (jeffy-mathew) [inline Makefile:42]: do we need to also think about injecting a global variable `Version` during linking?

Comment (haani-niyaz) [reply Makefile:42]: @jeffy-mathew yup, it is useful to have but perhaps as a separate issue. Feel free to create one.

File: Dockerfile [hunk 3]
```diff
@@ -1,13 +1,27 @@
-FROM golang:1.18 as build
-WORKDIR /go/src/app
+FROM golang:1.18.3-alpine3.16 as builder
```
[the remark this answers is not in the document]
Comment (mridulganga) [reply Dockerfile]: addressed and updated this.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (mridulganga) [reply Makefile:42]: Removed the GOOS variable

File: Dockerfile (hunk 1 above)
Comment (mridulganga) [reply Dockerfile]: Removed port

File: .gitignore [hunk 4]
```diff
@@ -15,3 +15,7 @@
 # vendor/
 
 bin/
+
+# code editor personal settings
+.vscode/
```
Comment (M1kep) [inline .gitignore:20]: Would it be beneficial to still keep task, launch, extensions etc? Personally have used this as a reference in the past: https://github.com/github/gitignore/blob/main/Global/VisualStudioCode.gitignore

Comment (mridulganga) [reply .gitignore:20]: we should not impose anyone's preferences of IDE config on others I think.

File: Dockerfile [hunk 5]
```diff
@@ -1,13 +1,23 @@
...
+# build executable
+RUN make build
 
 # Using a distroless image from https://github.com/GoogleContainerTools/distroless
-# Image sourced from https://console.cloud.google.com/gcr/images/distroless/global/static
 FROM gcr.io/distroless/static:nonroot
-COPY --from=build /go/bin/app /
+
+# copy executable from builder image to final images
+COPY --from=build /src/bin/app /
```
Comment (haani-niyaz) [inline Dockerfile:17]: @mridulganga this needs to be updated to the image built from the previous stage. You renamed it to `builder` but it is referenced as `build` here. Was this tested locally?

File: Dockerfile [hunk 6]
```diff
@@ -1,13 +1,23 @@
...
+
+# build executable
+RUN make build
 
 # Using a distroless image from https://github.com/GoogleContainerTools/distroless
-# Image sourced from https://console.cloud.google.com/gcr/images/distroless/global/static
 FROM gcr.io/distroless/static:nonroot
-COPY --from=build /go/bin/app /
+
+# copy executable from builder image to final images
```
Comment (haani-niyaz) [inline Dockerfile:16]: @mridulganga Can we follow the comment convention of using Uppercase first letter, thanks. 

Also I think it should say "Copy executable from builder image"

File: Makefile [hunk 7]
```diff
@@ -1,5 +1,7 @@
 SHELL=/bin/bash -e -o pipefail
```
Comment (haani-niyaz) [inline Makefile:1]: @mridulganga, Not sure if you tested this but this line would cause an error since `/bin/bash` doesn't exist in alpine. To make this work and for the sake of portability we should use the default shell `/bin/sh` so I would suggest removing this line.

File: Dockerfile (hunk 6 above)
Comment (mridulganga) [reply Dockerfile]: Updated comments to have first letter caps

File: Dockerfile (hunk 5 above)
Comment (mridulganga) [reply Dockerfile]: Done this change now.

File: Dockerfile (hunk 6 above)
Comment (haani-niyaz) [reply Dockerfile]: @mridulganga I don't know what you mean by "final images". The comment should be updated as I pointed out earlier.

File: .gitignore (hunk 4 above)
Comment (pallasite99) [inline .gitignore:20]: Forgive my persistence here, but why is this needed @mridulganga?

I think its quite a given that one should not include the `.vscode` folder here since that is used for local development.

File: sonar-project.properties [hunk 8]
```diff
@@ -10,3 +10,6 @@ sonar.organization=rog-golang-buddies
 
 # Encoding of the source code. Default is default system encoding
 #sonar.sourceEncoding=UTF-8
+
+# ensure that test files are identified and not used to check coverage
+sonar.test.inclusions=**/*_test.go
```
Comment (dcristobalh) [inline sonar-project.properties:15]: If we decide to remove this file, we could add an argument in gh sonarcloud workflow:
`-Dsonar.test.inclusions=**/*_test.go`

File: .gitignore (hunk 4 above)
Comment (mridulganga) [reply .gitignore:20]: Since .vscode is used for local dev, we should not checkin that folder, so I included the same in .gitignore.
